File: src/PythonClient/lib/quixstreaming/models/streamreader/streamparametersreader.py
from typing import List

from quixstreaming.eventhook import EventHook

from quixstreaming import __importnet, ParameterDefinition
import Quix.Sdk.Streaming

from quixstreaming.models.parameterdataraw import ParameterDataRaw
from quixstreaming.models.parameterdata import ParameterData
from quixstreaming.models.parametersbufferconfiguration import ParametersBufferConfiguration
from quixstreaming.models.streamreader.parametersbufferreader import ParametersBufferReader


class StreamParametersReader(object):

    def __init__(self, net_object: Quix.Sdk.Streaming.Models.StreamReader.StreamParametersReader):
        """
            Initializes a new instance of StreamParametersReader.
            NOTE: Do not initialize this class manually, use StreamReader.parameters to access an instance of it

            Parameters:

            net_object (.net object): The .net object representing a StreamParametersReader
        """
        if net_object is None:
            raise Exception("StreamParametersReader is none")
        self.__wrapped = net_object

        self.on_definitions_changed = EventHook(name="StreamParametersReader.on_definitions_changed")
        """
        Raised when the definitions have changed for the stream. Access "definitions" for latest set of parameter definitions 

        Has no arguments
        """

        def __on_definitions_changed_net_handler():
            self.on_definitions_changed.fire()
        self.__wrapped.OnDefinitionsChanged += __on_definitions_changed_net_handler



        def __on_read_net_handler(arg):
            self.on_read.fire(ParameterData(arg))

        def __on_first_sub():
            self.__wrapped.OnRead += __on_read_net_handler

        def __on_last_unsub():
            self.__wrapped.OnRead -= __on_read_net_handler

        self.on_read = EventHook(__on_first_sub, __on_last_unsub, name="StreamParametersReader.on_read")
        """
        Event raised when data is available to read (without buffering) 
        This event does not use Buffers and data will be raised as they arrive without any processing.

        Has one argument of type ParameterData
        """


        def __on_read_raw_net_handler(arg):
            converted = ParameterDataRaw(arg)
            self.on_read_raw.fire(converted)

        def __on_first_raw_sub():
            self.__wrapped.OnReadRaw += __on_read_raw_net_handler

        def __on_last_raw_unsub():
            self.__wrapped.OnReadRaw -= __on_read_raw_net_handler

        self.on_read_raw = EventHook(__on_first_raw_sub, __on_last_raw_unsub, name="StreamParametersReader.on_read_raw")
        """
        Event raised when data is available to read (without buffering) in raw transport format 
        This event does not use Buffers and data will be raised as they arrive without any processing. 

        Has one argument of type ParameterData in raw transport format 
        """



        def __on_read_pandas_net_handler(arg):
            converted = ParameterDataRaw(arg).to_panda_frame()
            self.on_read_pandas.fire(converted)

        def __on_first_raw_sub():
            self.__wrapped.OnReadRaw += __on_read_pandas_net_handler

        def __on_last_raw_unsub():
            self.__wrapped.OnReadRaw -= __on_read_pandas_net_handler

        self.on_read_pandas = EventHook(__on_first_raw_sub, __on_last_raw_unsub, name="StreamParametersReader.on_read_pandas")
        """
        Event raised when data is available to read (without buffering) in raw transport format 
        This event does not use Buffers and data will be raised as they arrive without any processing. 

        Has one argument of type ParameterData in raw transport format 
        """



    @property
    def definitions(self) -> List[ParameterDefinition]:
        """ Gets the latest set of parameter definitions """
        items = []
        for element in self.__wrapped.Definitions:
            item = ParameterDefinition(element)
            items.append(item)
        return items

    def create_buffer(self, *parameter_filter: str, buffer_configuration: ParametersBufferConfiguration = None) -> ParametersBufferReader:
        """
        Creates a new buffer for reading data according to the provided parameter_filter and buffer_configuration
        :param parameter_filter: 0 or more parameter identifier to filter as a whitelist. If provided, only these
            parameters will be available through this buffer
        :param buffer_configuration: an optional ParameterBufferConfiguration.

        :returns: a ParametersBufferReader which will raise new parameters read via .on_read event
        """

        actual_filters = []
        for param_filter in parameter_filter:
            if isinstance(param_filter, ParametersBufferConfiguration):
                buffer_configuration = param_filter
                break
            actual_filters.append(param_filter)

        buffer_config = None
        if buffer_configuration is not None:
            buffer_config = buffer_configuration.convert_to_net()
        dotnet_obj = self.__wrapped.CreateBuffer(buffer_config, actual_filters)
        return ParametersBufferReader(dotnet_obj)

File: src/CsharpClient/Quix.Sdk.Streaming/Models/StreamReader/StreamParametersReader.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Quix.Sdk.Process.Models;

namespace Quix.Sdk.Streaming.Models.StreamReader
{
    /// <summary>
    /// Helper class for reader <see cref="ParameterDefinitions"/> and <see cref="TimeseriesData"/>
    /// </summary>
    public class StreamParametersReader : IDisposable
    {
        private readonly IInputTopic topic;
        private readonly IStreamReaderInternal streamReader;

        /// <summary>
        /// Initializes a new instance of <see cref="StreamParametersReader"/>
        /// </summary>
        /// <param name="topic">The topic the stream to what this reader belongs to</param>
        /// <param name="streamReader">Stream reader owner</param>
        internal StreamParametersReader(IInputTopic topic, IStreamReaderInternal streamReader)
        {
            this.topic = topic;
            this.streamReader = streamReader;

            this.streamReader.OnParameterDefinitionsChanged += OnParameterDefinitionsChangedEventHandler;

            this.streamReader.OnTimeseriesData += OnTimeseriesDataEventHandler;
            this.streamReader.OnTimeseriesData += OnTimeseriesDataRawEventHandler;
        }

        private void OnParameterDefinitionsChangedEventHandler(IStreamReader sender, ParameterDefinitions parameterDefinitions)
        {
            this.LoadFromProcessDefinitions(parameterDefinitions);

            this.OnDefinitionsChanged?.Invoke(this.streamReader, new ParameterDefinitionsChangedEventArgs(this.topic, this.streamReader));
        }

        /// <summary>
        /// Create a new Parameters buffer for reading data
        /// </summary>
        /// <param name="bufferConfiguration">Configuration of the buffer</param>
        /// <param name="parametersFilter">List of parameters to filter</param>
        /// <returns>Parameters reading buffer</returns>
        public TimeseriesBufferReader CreateBuffer(TimeseriesBufferConfiguration bufferConfiguration = null, params string[] parametersFilter)
        {
            var buffer = new TimeseriesBufferReader(this.topic, this.streamReader, bufferConfiguration, parametersFilter);
            this.Buffers.Add(buffer);

            return buffer;
        }

        /// <summary>
        /// Create a new Parameters buffer for reading data
        /// </summary>
        /// <param name="parametersFilter">List of parameters to filter</param>
        /// <returns>Parameters reading buffer</returns>
        public TimeseriesBufferReader CreateBuffer(params string[] parametersFilter)
        {
            var buffer = new TimeseriesBufferReader(this.topic, this.streamReader, null, parametersFilter);
            this.Buffers.Add(buffer);

            return buffer;
        }

        /// <summary>
        /// Raised when the parameter definitions have changed for the stream.
        /// See <see cref="Definitions"/> for the latest set of parameter definitions
        /// </summary>
        public event EventHandler<ParameterDefinitionsChangedEventArgs> OnDefinitionsChanged;

        /// <summary>
        /// Event raised when data is available to read (without buffering)
        /// This event does not use Buffers and data will be raised as they arrive without any processing.
        /// </summary>
        public event EventHandler<TimeseriesDataReadEventArgs> OnRead;

        /// <summary>
        /// Event raised when data is available to read (without buffering) in raw transport format
        /// This event does not use Buffers and data will be raised as they arrive without any processing.
        /// </summary>
        public event EventHandler<TimeseriesDataRawReadEventArgs> OnReadRaw;

        /// <summary>
        /// Gets the latest set of event definitions
        /// </summary>
        public List<ParameterDefinition> Definitions { get; private set; } = new List<ParameterDefinition>();

        /// <summary>
        /// List of buffers created for this stream
        /// </summary>
        internal List<TimeseriesBufferReader> Buffers { get; private set; } = new List<TimeseriesBufferReader>();

        private void LoadFromProcessDefinitions(Process.Models.ParameterDefinitions definitions)
        {
            var defs = new List<ParameterDefinition>();
            
            if (definitions.Parameters != null) 
                this.ConvertParameterDefinitions(definitions.Parameters, "").ForEach(d => defs.Add(d));
            if (definitions.ParameterGroups != null)
                this.ConvertGroupParameterDefinitions(definitions.ParameterGroups, "").ForEach(d => defs.Add(d));

            this.Definitions = defs;
        }

        private List<ParameterDefinition> ConvertParameterDefinitions(List<Process.Models.ParameterDefinition> parameterDefinitions, string location)
        {
            var result = parameterDefinitions.Select(d => new ParameterDefinition
            {
                Id = d.Id,
                Name = d.Name,
                Description = d.Description,
                MinimumValue = d.MinimumValue,
                MaximumValue = d.MaximumValue,
                Unit = d.Unit,
                Format = d.Format,
                CustomProperties = d.CustomProperties,
                Location = location
            }).ToList();

            return result;
        }

        private List<ParameterDefinition> ConvertGroupParameterDefinitions(List<Process.Models.ParameterGroupDefinition> parameterGroupDefinitions, string location)
        {
            var result = new List<ParameterDefinition>();

            foreach (var group in parameterGroupDefinitions)
            {
                if (group.Parameters != null)
                    this.ConvertParameterDefinitions(group.Parameters, location + "/" + group.Name).ForEach(d => result.Add(d));
                if (group.ChildGroups != null)
                    this.ConvertGroupParameterDefinitions(group.ChildGroups, location + "/" + group.Name).ForEach(d => result.Add(d));
            }

            return result;
        }

        private void OnTimeseriesDataEventHandler(IStreamReader streamReader, Process.Models.TimeseriesDataRaw timeseriesDataRaw)
        {
            var tsdata = new TimeseriesData(timeseriesDataRaw, null, false, false);
            this.OnRead?.Invoke(streamReader, new TimeseriesDataReadEventArgs(this.topic, streamReader, tsdata));
        }

        private void OnTimeseriesDataRawEventHandler(IStreamReader streamReader, Process.Models.TimeseriesDataRaw timeseriesDataRaw)
        {
            this.OnReadRaw?.Invoke(streamReader, new TimeseriesDataRawReadEventArgs(this.topic, streamReader, timeseriesDataRaw));
        }

        /// <inheritdoc/>
        public void Dispose()
        {
            this.streamReader.OnParameterDefinitionsChanged -= OnParameterDefinitionsChangedEventHandler;
            this.streamReader.OnTimeseriesData -= OnTimeseriesDataEventHandler;
            this.streamReader.OnTimeseriesData -= OnTimeseriesDataRawEventHandler;
        }
    }

    public class ParameterDefinitionsChangedEventArgs
    {
        public ParameterDefinitionsChangedEventArgs(IInputTopic topic, IStreamReader reader)
        {
            this.Topic = topic;
            this.Stream = reader;
        }

        public IInputTopic Topic { get; }
        public IStreamReader Stream { get; }
    }

    public class TimeseriesDataReadEventArgs
    {
        public TimeseriesDataReadEventArgs(object topic, object stream, TimeseriesData data)
        {
            this.Topic = topic;
            this.Stream = stream;
            this.Data = data;
        }
        
        /// <summary>
        /// Topic of type <see cref="IInputTopic"/> or <see cref="IOutputTopic"/>
        /// </summary>
        public object Topic { get; }
        
        /// <summary>
        /// Stream of type <see cref="IStreamReader"/> or <see cref="IStreamWriter"/>
        /// </summary>
        public object Stream { get; }
        
        public TimeseriesData Data { get; }
    }
    
    public class TimeseriesDataRawReadEventArgs
    {
        public TimeseriesDataRawReadEventArgs(object topic, object stream, TimeseriesDataRaw data)
        {
            this.Topic = topic;
            this.Stream = stream;
            this.Data = data;
        }

        /// <summary>
        /// Topic of type <see cref="IInputTopic"/> or <see cref="IOutputTopic"/>
        /// </summary>
        public object Topic { get; }
        
        /// <summary>
        /// Stream of type <see cref="IStreamReader"/> or <see cref="IStreamWriter"/>
        /// </summary>
        public object Stream { get; }
        
        public TimeseriesDataRaw Data { get; }
    }
}
